### PR TITLE
Improvements in the class based model

### PR DESCRIPTION
**Pull request by zackliu**
Some improvements to make class based model much closer to SignalR SDK. Also make it more convenient.
* Add a more convenient negotiate approach which can assign claims in runtime.
* Use `OnConnected` and `OnDisconnected` to match that in SDK.
* Extract a `SignalRFilterAttribute` for easier way to implement a customer filter.

Also, design a much strict convention that class based model must use non-arg constructor.

### CODE REVIEW

File: src/SignalRServiceExtension/TriggerBindings/ServerlessHub.cs [hunk 1]
```diff
@@ -30,6 +38,32 @@ public ServerlessHub()
 
         public IUserGroupManager UserGroups { get; }
 
+        public string HubName { get; }
```
Comment (JialinXin) [inline src/SignalRServiceExtension/TriggerBindings/ServerlessHub.cs]: Add summary for public properties.

File: samples/bidirectional-chat/csharp/Function.cs [hunk 2]
```diff
@@ -80,8 +78,8 @@ public async Task LeaveUserFromGroup([SignalRTrigger]InvocationContext invocatio
             await UserGroups.RemoveFromGroupAsync(userName, groupName);
         }
 
-        [FunctionName(nameof(Disconnect))]
-        public void Disconnect([SignalRTrigger]InvocationContext invocationContext)
+        [FunctionName(nameof(OnDisconnect))]
+        public void OnDisconnect([SignalRTrigger]InvocationContext invocationContext)
```
Comment (wanlwanl) [inline samples/bidirectional-chat/csharp/Function.cs]: Why not OnDisconnected...? it doesn't consistent to `OnConnected`

File: src/SignalRServiceExtension/TriggerBindings/InvocationContextExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Azure.SignalR.Management;
+
+namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
+{
+    public static class InvocationContextExtensions
```
Comment (wanlwanl) [inline src/SignalRServiceExtension/TriggerBindings/InvocationContextExtensions.cs]: add summary?

File: src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            }
+            // Should not hit the Exception.
+            throw new InvalidOperationException($"{nameof(FunctionExceptionContext)} doesn't contain {nameof(InvocationContext)}.");
+        }
+
+        /// <summary>
+        /// Executed before the Function method being executed.
+        /// Throwing exceptions can terminate the Function execution and response the invocation failure.
+        /// </summary>
+        public abstract Task FilterAsync(InvocationContext invocationContext, CancellationToken cancellationToken);
```
Comment (wanlwanl) [inline src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs]: Can we make the filter returns true/false?

File: samples/bidirectional-chat/csharp/Function.cs (hunk 2 above)
Comment (zackliu) [reply samples/bidirectional-chat/csharp/Function.cs]: Oh, nice catch.

File: src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs [hunk 5]
```diff
@@ -62,22 +62,29 @@ internal SignalRTriggerAttribute GetParameterResolvedAttribute(SignalRTriggerAtt
             if (declaredType != null && declaredType.IsSubclassOf(typeof(ServerlessHub)))
             {
                 // Class based model
+                if (!string.IsNullOrEmpty(hubName) ||
```
Comment (wanlwanl) [inline src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs]: make the properties' setters access level less than public?

File: src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs [hunk 6]
```diff
@@ -125,15 +132,32 @@ private void ValidateSignalRTriggerAttributeBinding(SignalRTriggerAttribute attr
 
         private string GetCategoryFromMethodName(string name)
         {
-            if (string.Equals(name, Event.Connected, StringComparison.OrdinalIgnoreCase) ||
-                string.Equals(name, Event.Disconnected, StringComparison.OrdinalIgnoreCase))
+            if (string.Equals(name, Constants.OnConnected, StringComparison.OrdinalIgnoreCase) ||
+                string.Equals(name, Constants.OnDisconnected, StringComparison.OrdinalIgnoreCase))
```
Comment (JialinXin) [inline src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs]: hard code map to function name might be easily break and hard for customer to identify problem. Any way to notice the method name mapping logic?

File: src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs (hunk 5 above)
Comment (wanlwanl) [reply src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs]: or move them to another class?

File: src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs (hunk 4 above)
Comment (zackliu) [reply src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs]: The message of exception will be sent back to client side. So, I think let customer to throw exception maybe better. Or if they returns false, what should we throw?

File: src/SignalRServiceExtension/TriggerBindings/InvocationContextExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        public static async Task<IHubClients> GetClientsAsync(this InvocationContext invocationContext)
```
Comment (wanlwanl) [inline src/SignalRServiceExtension/TriggerBindings/InvocationContextExtensions.cs]: possilble to put Clients/Groups/UserGroups to InvocationContext?
Looks the Clients/Groups/UserGroups are only for C#, maybe add a new output of SignalRTriggerAttri that's only for C#?
Then the interface will be much similar to IHubContext

Comment (wanlwanl) [reply src/SignalRServiceExtension/TriggerBindings/InvocationContextExtensions.cs]: Or add a new class implements IHubContext and use the InvocationContext as ctor param?

Comment (zackliu) [reply src/SignalRServiceExtension/TriggerBindings/InvocationContextExtensions.cs]: I think in most C# cases, they should use class based model. Then they use that just like what they do in SDK. And for other few customers, we provided an extension to let them use it easily. It's not a common case, so as simple as this is OK I think.
Put them to InvocationContext is not good, and InvocationContext should keep simple as it passes between multiple language.

File: src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs (hunk 5 above)
Comment (zackliu) [reply src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs]: Removing Setters for these properties is good.

File: src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs (hunk 4 above)
Comment (wanlwanl) [reply src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs]: I remember the exception throwed in function will respond 500 internal server error to client, so we can throw in OnExecutingAsync?

Comment (zackliu) [reply src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs]: No, it the behavior of httpTrigger. In SignalRTrigger, we handle the exception and send back the invocation response with message.

File: samples/bidirectional-chat/csharp/Function.cs [hunk 8]
```diff
@@ -80,8 +78,8 @@ public async Task LeaveUserFromGroup([SignalRTrigger]InvocationContext invocatio
             await UserGroups.RemoveFromGroupAsync(userName, groupName);
         }
 
-        [FunctionName(nameof(Disconnect))]
-        public void Disconnect([SignalRTrigger]InvocationContext invocationContext)
+        [FunctionName(nameof(OnDisconnected))]
+        public void OnDisconnected([SignalRTrigger]InvocationContext invocationContext)
```
Comment (vicancy) [inline samples/bidirectional-chat/csharp/Function.cs]: is `OnDisconnectedAsync` a better matching? why `OnConnected` is Task `OnDisconnected` is void?

File: src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
+{
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = true)]
+#pragma warning disable CS0618 // Type or member is obsolete
+    public abstract class SignalRFilterAttribute : FunctionInvocationFilterAttribute
+    {
+        public override Task OnExecutingAsync(FunctionExecutingContext executingContext,
+            CancellationToken cancellationToken)
+        {
+            if (executingContext.Arguments.FirstOrDefault().Value is InvocationContext invocationContext)
```
Comment (vicancy) [inline src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs]: `FirstOrDefault()` can be `null`?

Comment (zackliu) [reply src/SignalRServiceExtension/TriggerBindings/SignalRFilterAttribute.cs]: It's a `KeyValuePair` and won't be null

File: samples/bidirectional-chat/csharp/Function.cs (hunk 8 above)
Comment (zackliu) [reply samples/bidirectional-chat/csharp/Function.cs]: It's decided by customer, and we don't care about the return value. And the name need to be a convention, so I think the one without `Async` is enough.

File: src/SignalRServiceExtension/TriggerBindings/ServerlessHub.cs [hunk 10]
```diff
@@ -2,34 +2,88 @@
...
+using System.Security.Claims;
 using Microsoft.AspNetCore.SignalR;
 using Microsoft.Azure.SignalR.Management;
 
 namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
 {
     /// <summary>
     /// When a class derived from <see cref="ServerlessHub"/>,
-    /// all the method in the are identified as using class based model.
+    /// all the method in the class are identified as using class based model.
```
Comment (JialinXin) [inline src/SignalRServiceExtension/TriggerBindings/ServerlessHub.cs]: typo: methods

File: samples/bidirectional-chat/csharp/Function.cs [hunk 11]
```diff
@@ -12,25 +12,23 @@ namespace FunctionApp
...
-            [SignalRConnectionInfo(HubName = Hub, UserId = "{headers.x-ms-signalr-user-id}")] SignalRConnectionInfo connectionInfo)
+        public SignalRConnectionInfo Negotiate([HttpTrigger(AuthorizationLevel.Anonymous)]HttpRequest req)
         {
-            return connectionInfo;
+            return Negotiate(req.Headers["x-ms-signalr-user-id"], GetClaims(req.Headers["Authorization"]));
         }
 
-        [FunctionName(nameof(Connected))]
-        public async Task Connected([SignalRTrigger]InvocationContext invocationContext, ILogger logger)
+        [FunctionName(nameof(OnConnected))]
```
Comment (vwxyzh) [inline samples/bidirectional-chat/csharp/Function.cs]: function name should be constant

File: src/SignalRServiceExtension/TriggerBindings/ServerlessHub.cs [hunk 12]
```diff
@@ -2,34 +2,88 @@
...
         private bool _disposed;
+        private readonly IServiceManager _serviceManager;
 
         public ServerlessHub()
         {
-            var hubName = GetType().Name;
-            var hubContext = StaticServiceHubContextStore.Get().GetAsync(hubName).GetAwaiter().GetResult();
+            HubName = GetType().Name;
+            var store = StaticServiceHubContextStore.Get();
+            var hubContext = store.GetAsync(HubName).GetAwaiter().GetResult();
```
Comment (vwxyzh) [inline src/SignalRServiceExtension/TriggerBindings/ServerlessHub.cs]: emm....

File: samples/bidirectional-chat/csharp/Function.cs (hunk 11 above)
Comment (zackliu) [reply samples/bidirectional-chat/csharp/Function.cs]: Method name in class based model has special meanings. When method name changed, function name changes make sense. Constant is OK but `nameof` is also acceptable.

File: samples/bidirectional-chat/csharp/Function.cs [hunk 13]
```diff
@@ -12,25 +12,23 @@ namespace FunctionApp
...
+        public async Task OnConnected([SignalRTrigger]InvocationContext invocationContext, ILogger logger)
```
Comment (chenkennt) [inline samples/bidirectional-chat/csharp/Function.cs]: How add `OnConnected` and `OnDisconnected` as virtual methods in base class?

Comment (zackliu) [reply samples/bidirectional-chat/csharp/Function.cs]: We don't know parameters of these two methods. Just like `Configure` and `ConfigureService` in `Startup`, which can only restrict by convention.

File: src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs [hunk 14]
```diff
@@ -62,22 +62,29 @@ internal SignalRTriggerAttribute GetParameterResolvedAttribute(SignalRTriggerAtt
             if (declaredType != null && declaredType.IsSubclassOf(typeof(ServerlessHub)))
             {
                 // Class based model
+                if (!string.IsNullOrEmpty(hubName) ||
+                    !string.IsNullOrEmpty(category) ||
+                    !string.IsNullOrEmpty(@event) ||
+                    parameterNames.Length != 0)
+                {
+                    throw new ArgumentException($"{nameof(SignalRTriggerAttribute)} must use No-Arg Constructor in class based model.");
```
Comment (chenkennt) [inline src/SignalRServiceExtension/TriggerBindings/SignalRTriggerBindingProvider.cs]: In C# it's officially called *parameterless constructor*?

**Pull request merged**